### PR TITLE
Support pre-normalization for FT Encoder fp32

### PR DESCRIPTION
**Pull request by tianxin1860**
### PR types

New features
### PR changes

Models
### Description

1. support pre-normalization for FT Encoder fp32

**Notes:**

1. FT encoder sample will compile fp32 and fp16, this pr only support fp32 pre-normalizaiton, so comment `add_subdirectory(sample)` in `FasterTransformer/CMakeLists.txt`
2. jit compile only support Paddle 2.1.x, 2.0.x will throw error.

**TODO:**
- [ ] support fp16 post-normalization
- [ ] support int8 post-normalization

### CODE REVIEW

File: paddlenlp/ops/patches/FasterTransformer/CMakeLists.txt [hunk 1]
```diff
@@ -149,4 +149,4 @@ link_directories(
 
 add_subdirectory(fastertransformer)
 add_subdirectory(tools)
-add_subdirectory(sample)
+#add_subdirectory(sample)
```
Comment (FrostML) [inline paddlenlp/ops/patches/FasterTransformer/CMakeLists.txt:226]: 这里注释了是为了解决什么问题么？

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 2]
```diff
@@ -73,4 +73,153 @@ void update_logits_v2(float* logits,
...
+      (float)(out[blockIdx.x * n + tid] + bias_and_input[blockIdx.x * n + tid]);
+
+  out[blockIdx.x * n + tid] = local_out;
+}
+
+// Encoder kernels
+template <typename T>
+__global__ void encoder_layernorm(
+    T* out, const T* input, const T* gamma, const T* beta, int n) {
+  int tid = threadIdx.x;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu:142]: layer_norm实现最好参考这里 https://github.com/NVIDIA/FasterTransformer/blob/v3.1/fastertransformer/cuda/open_decoder.cu#L1621 或 https://github.com/NVIDIA/FasterTransformer/blob/v4.0/fastertransformer/cuda/transformer_kernels.cu#L517 搞成general版本的，这里的size会受些限制

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 3]
```diff
@@ -73,4 +73,153 @@ void update_logits_v2(float* logits,
...
+                                             int n) {
+  int tid = threadIdx.x;
+
+  __shared__ float s_mean;
+  __shared__ float s_variance;
+  float mean = 0.0f;
+  float variance = 0.0f;
+
+  float local_out = 0.0f;
+  local_out += (float)(out[blockIdx.x * n + tid] + input[blockIdx.x * n + tid] +
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 和encoder_layernorm一样最好是搞成支持任意size的general版本

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 4]
```diff
@@ -73,4 +73,153 @@ void update_logits_v2(float* logits,
...
+__global__ void add_bias_input(T* out,
+                               const T* bias_and_input,
+                               const T* bias,
+                               int n) {
+  int tid = threadIdx.x;
+
+  __shared__ float s_mean;
+  __shared__ float s_variance;
+  float mean = 0.0f;
+  float variance = 0.0f;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: s_mean、s_variance这里没用的话去掉吧

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu (hunk 2 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu:142]: 已修正，thanks

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu (hunk 3 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 已修正，thanks

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu (hunk 4 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 已修正，thanks

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 5]
```diff
@@ -73,4 +73,178 @@ void update_logits_v2(float* logits,
...
+                                                 const T* input,
+                                                 const T* bias,
+                                                 const T* gamma,
+                                                 const T* beta,
+                                                 int m,
+                                                 int n,
+                                                 cudaStream_t stream) {
+  dim3 grid(m);
+  dim3 block(n);
+  assert(n <= 1024);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 下面的kernel支持的size是generalize的话这里的assert去掉吧

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 6]
```diff
@@ -73,4 +73,178 @@ void update_logits_v2(float* logits,
...
+template <typename T>
+void add_bias_input_kernelLauncher(T* out,
+                                   const T* bias_and_input,
+                                   const T* bias,
+                                   int m,
+                                   int n,
+                                   cudaStream_t stream) {
+  dim3 grid(m);
+  dim3 block(n);
+  assert(n <= 1024);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 同上，下面的kernel支持的size是generalize的话这里的assert去掉吧

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu [hunk 7]
```diff
@@ -73,4 +73,178 @@ void update_logits_v2(float* logits,
...
+void layernorm_kernelLauncher(T* out,
+                              const T* input,
+                              const T* gamma,
+                              const T* beta,
+                              int m,
+                              int n,
+                              cudaStream_t stream) {
+  dim3 grid(m);
+  dim3 block(n);
+  assert(n <= 1024);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 同上，下面的kernel支持的size是generalize的话这里的assert去掉吧

Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 已修正，thanks

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu (hunk 6 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 已修正，thanks

File: paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu (hunk 5 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/cuda/cuda_kernels.cu]: 已修正，thanks

File: paddlenlp/ops/CMakeLists.txt [hunk 8]
```diff
@@ -147,6 +147,9 @@ file(TO_NATIVE_PATH ${CMAKE_BINARY_DIR}/${THIRD_PATH}/source/${THIRD_PARTY_NAME}
 file(TO_NATIVE_PATH ${OPS_SOURCE_DIR}/patches/FasterTransformer/cuda/transformer_decoder.cu trans_decoder_cu_src)
 file(TO_NATIVE_PATH ${OPS_SOURCE_DIR}/patches/FasterTransformer/transformer_decoder.h trans_decoder_h_src)
 
+file(TO_NATIVE_PATH ${OPS_SOURCE_DIR}/patches/FasterTransformer/bert_encoder_transformer.h bert_encoder_transformer_h_src)
```
Comment (FrostML) [inline paddlenlp/ops/CMakeLists.txt]: 这句，在 174 行已经有了

File: paddlenlp/ops/patches/FasterTransformer/CMakeLists.txt (hunk 1 above)
Comment (tianxin1860) [reply paddlenlp/ops/patches/FasterTransformer/CMakeLists.txt:226]: FT sample 编译会默认编译 fp16 版本，目前 Pre-normalization 实现的 kernel 只支持 fp32, fp16 后续支持。

File: paddlenlp/ops/CMakeLists.txt (hunk 8 above)
Comment (tianxin1860) [reply paddlenlp/ops/CMakeLists.txt]: 已修正，thanks~

File: paddlenlp/ops/faster_transformer/transformer/encoder.py [hunk 9]
```diff
@@ -173,7 +175,8 @@ def encoder_layer_forward(self,
         # amax_list=paddle.to_tensor([]),  # int8 mode is not supported.
         n_head=self._config['nhead'],
         size_per_head=self._config['d_model'] // self._config['nhead'],
-        is_gelu=self._config['activation'] == 'gelu')
+        is_gelu=self._config['activation'] == 'gelu',
+        normalize_before=self._config['normalize_before'] == True)
     return src
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/transformer/encoder.py:182]: FT encoder这里支持的src_mask输入的形状和nn.TransformerEncoder的不一样，这里写的文档有问题 @LiuChiachi  还请注意

Comment (LiuChiachi) [reply paddlenlp/ops/faster_transformer/transformer/encoder.py:182]: 文档已经修改在#1029

**Review by guoshengCS — APPROVED** [review]
